packages/bff/src/utilities/samlValidator.ts

```diff
@@ -0,0 +1,165 @@
+  }
+  const atLeastOneEnvelopedSignature = transforms.some(
+    (t) => t.Algorithm === TRANSFORM_ENVELOPED_SIGNATURE
+  );
+  const allAlgorithmsAreValid = transforms.every(
+    (t) =>
+      t.Algorithm &&
+      [
+        TRANSFORM_ENVELOPED_SIGNATURE,
+        TRANSFORM_C14N_EXCL_OMIT_COMMENTS,
+        TRANSFORM_C14N_EXCL_WITH_COMMENTS,
+      ].includes(t.Algorithm)
+  );
+  if (!atLeastOneEnvelopedSignature || !allAlgorithmsAreValid) {
+    throw samlNotValid("Transforms are not compliant");
+  }
+};
+
+export const validateSamlResponse = (samlResponse: string): SAMLResponse => {
+  const xml = new XMLParser({
+    ignoreDeclaration: true,
+    removeNSPrefix: true,
+    ignoreAttributes: false,
+    attributeNamePrefix: "",
+    isArray: (name) =>
+      [
+        "Assertion",
+        "AudienceRestriction",
+        "Audience",
+        "AttributeValue",
+        "Transform",
+      ].indexOf(name) !== -1,
+  }).parse(samlResponse);
+
+  const { success, data: saml, error } = SAMLResponse.safeParse(xml);
+
+  if (!success) {
+    throw samlNotValid(error.message);
+  }
+
+  if (!saml.Response) {
+    throw samlNotValid("Response not found");
+  }
+  const response = saml.Response;
+  if (!response.Signature) {
+    throw samlNotValid("Missing Signature");
+  }
+  if (!response.Assertion || response.Assertion.length === 0) {
+    throw samlNotValid("Missing Assertions");
+  }
+  const assertions = response.Assertion;
+  const conditions = assertions
+    .flatMap((a) => a.Conditions)
+    .filter(filterUndefined);
+  const audienceRestrictions = conditions
+    .flatMap((c) => c.AudienceRestriction)
+    .filter(filterUndefined);
+  if (audienceRestrictions.length === 0) {
+    throw samlNotValid("Missing Audience Restriction");
+  }
+  const notBeforeConditions = conditions
+    .map((c) => c.NotBefore)
+    .filter(filterUndefined);
+  if (notBeforeConditions.length === 0) {
+    throw samlNotValid("Missing Not Before Restrictions");
+  }
+  const notOnOrAfterConditions = conditions
+    .map((c) => c.NotOnOrAfter)
+    .filter(filterUndefined);
+  if (notOnOrAfterConditions.length === 0) {
+    throw samlNotValid("Missing Not On Or After Restrictions");
+  }
+  const attributeStatements = assertions
+    .flatMap((a) => a.AttributeStatement)
+    .filter(filterUndefined);
+  if (attributeStatements.length === 0) {
+    throw samlNotValid("Missing Attribute Statement");
+  }
+  const attributes = attributeStatements
+    .flatMap((a) => a.Attribute)
+    .filter(filterUndefined);
+  if (attributes.length === 0) {
+    throw samlNotValid("Missing Attributes");
+  }
+  const now = +new Date();
+
+  validateSignature(saml);
+
+  if (!notBeforeConditions.every((nb) => now > +new Date(nb))) {
+    throw samlNotValid("Conditions notbefore are not compliant");
+  }
+  if (!notOnOrAfterConditions.every((noa) => now < +new Date(noa))) {
+    throw samlNotValid("Conditions NotOnOrAfter are not compliant");
+  }
```

Comment:
There should be no need to manually coerce the dates as number with `+` since it's already done implicitly in the comparison expressions. I would remove it to avoid confusion

```suggestion
  const now = new Date();

  validateSignature(saml);

  if (!notBeforeConditions.every((nb) => now > new Date(nb))) {
    throw samlNotValid("Conditions notbefore are not compliant");
  }
  if (!notOnOrAfterConditions.every((noa) => now < new Date(noa))) {
    throw samlNotValid("Conditions NotOnOrAfter are not compliant");
  }
```